haicrypt/hc_nettle_aes.c

```diff
@@ -0,0 +1,448 @@
+     */
+    if (nbout_p != NULL)
+      *nbout_p = 0;
+    return (-1);
+  }
+  return (0);
+}
+
+
+
+static int
+hcNettle_AES_Decrypt (hcrypt_CipherData * cipher_data, hcrypt_Ctx * ctx,
+    hcrypt_DataDesc * in_data, int nbin, void *out_p[], size_t out_len_p[],
+    int *nbout_p)
+{
+  hcNettle_AES_data *aes_data = (hcNettle_AES_data *) cipher_data;
+  unsigned char *out_txt;
+  int out_len;
+  int iret = 0;
+
+  ASSERT (NULL != aes_data);
+  ASSERT (NULL != ctx);
+  ASSERT ((NULL != in_data) || (1 == nbin));    //Only one in_data[] supported
+
+  /* Reserve output buffer (w/no header) */
+  out_txt = hcNettle_AES_GetOutbuf (aes_data, 0, in_data[0].len);
+
+  if (NULL != out_txt) {
+    switch (ctx->mode) {
+      case HCRYPT_CTX_MODE_AESCTR:
+      {
+        /* Get current key (odd|even) from context */
+        AES_KEY *aes_key = &aes_data->aes_key[hcryptCtx_GetKeyIndex (ctx)];
+        unsigned char ctr[AES_BLOCK_SIZE];
+        unsigned char iv[AES_BLOCK_SIZE];
+
+        /* Get input message index (in network order) */
+        hcrypt_Pki pki = hcryptMsg_GetPki (ctx->msg_info, in_data[0].pfx, 1);
+
+        /*
+         * Compute the Initial Vector
+         * IV (128-bit):
+         *    0   1   2   3   4   5  6   7   8   9   10  11  12  13  14  15
+         * +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+         * |                   0s                  |      pki      |  ctr  |
+         * +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+         *                            XOR                         
+         * +---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+         * |                         nonce                         +
+         * +---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+         *
+         * pki    (32-bit): packet index
+         * ctr    (16-bit): block counter
+         * nonce (112-bit): number used once (salt)
+         */
+        memset (&ctr[0], 0, sizeof (ctr));
+        hcrypt_SetCtrIV ((unsigned char *) &pki, ctx->salt, iv);
+
+        /* Decrypt message (same as encrypt for CTR mode) */
+        ctr_crypt (aes_key,        /* ctx */
+                   aes128_encrypt, /* nettle_cipher_func */
+                   AES_BLOCK_SIZE, /* cipher blocksize */
+                   ctr,            /* ctr */
+                   in_data[0].len, /* length */
+                   out_txt,             /* dest */
+                   in_data[0].payload   /* src */);
+
+        out_len = in_data[0].len;
+        break;
+      }
+      case HCRYPT_CTX_MODE_AESECB:
+      {
+        int i;
+        int nb = in_data[0].len / AES_BLOCK_SIZE;
+        unsigned nbpad = ctx->msg_info->getPki (in_data[0].pfx, 0);     //Patch
+        AES_KEY *aes_key = &aes_data->aes_key[hcryptCtx_GetKeyIndex (ctx)];
+
+        /* Decrypt message (same as encrypt for CTR mode) */
+        for (i = 0; i < nb; i++) {
+          aes128_encrypt (aes_key, AES_BLOCK_SIZE, 
```

Comment:
I am not sure why this and the above should be `aes128_encrypt`, but this works for me.